airbyte-cdk/python/airbyte_cdk/sources/availability_strategy.py

```diff
@@ -0,0 +1,97 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+from abc import ABC, abstractmethod
+from requests import HTTPError
+from typing import Dict, List, Optional, Text, Tuple
+
+from airbyte_cdk.models.airbyte_protocol import SyncMode
+from airbyte_cdk.sources.streams import Stream
+
+
+class AvailabilityStrategy(ABC):
+    """
+    Abstract base class for checking stream availability
+    """
+
+    @abstractmethod
+    def check_availability(self, stream: Stream) -> Tuple[bool, any]:
+        """
+        Checks stream availability.
+
+        :param stream: stream
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available. Otherwise, the stream is unavailable for some reason and
+          the str should describe what went wrong.
+        """
+
+
+class HTTPAvailabilityStrategy(AvailabilityStrategy):
+    def check_availability(self, stream: Stream) -> Tuple[bool, str]:
+        """
+        Check stream availability by attempting to read the first record of the
+        stream.
+
+        :param stream: stream
+        :return: A tuple of (boolean, str). If boolean is true, then the stream
+          is available. Otherwise, the stream is unavailable for some reason and
+          the str should describe what went wrong.
+        """
+        try:
+            # Some streams need a stream slice to read records (e.g. if they have a SubstreamSlicer)
+            stream_slice = self._get_stream_slice(stream)
+            records = stream.read_records(sync_mode=SyncMode.full_refresh, stream_slice=stream_slice)
+            next(records)
+        except HTTPError as error:
+            return False, repr(error)
+        return True, None
+
+    def _get_stream_slice(self, stream):
+        # We wrap the return output of stream_slices() because some implementations return types that are iterable,
+        # but not iterators such as lists or tuples
+        slices = iter(
+            stream.stream_slices(
+                cursor_field=stream.cursor_field,
+                sync_mode=SyncMode.full_refresh,
+            )
+        )
+        try:
+            return next(slices)
+        except StopIteration:
+            return {}
+
+
+class ScopedAvailabilityStrategy(AvailabilityStrategy):
```

Comment:
The plan is to remove this before merging and keep it on another branch. As we review PRs for connectors implementing availabilitystrategies, if we see enough common functionality, we can make a specific availabilitystrategy for scoped availability.

I'll keep this here while still working though, so that it gets refactored along with anything else so it's easier to slot back in if necessary.